### PR TITLE
PUBDEV-5394 - h2o parser wrongly treats Ctr-M character as end of lin…

### PR DESCRIPTION
**Pull request by Pscheidl**
…e on linux

[JIRA PUBDEV-5394](https://0xdata.atlassian.net/browse/PUBDEV-5394)

### CODE REVIEW

**Comment by Pscheidl** [comment]
Seems to be fixed by counting quotes in two phases:

- Guess setup
- ParseChunk

This is still WIP - a test is required. A manual test on the data looks ok. Let's see how the tests go.
snimek obrazovky porizeny 2018-03-20 12-28-53

**Comment by michalkurka** [comment]
The idea makes sense to me Pavel! We should extend our test suite and possible re-use some tests from data.table

**Comment by Pscheidl** [comment]
I added one test of my own. Also made some additional fixes and optimizations. Added a previously deleted (by mistake) check for a mischmatched quote. 

Right now, I am trying to solve a problem with offset.

**Comment by Pscheidl** [comment]
Ready to review @michalkurka . Previous tests were all green. In the very last commit, I only fixed an offset leak for multi-line cells, causing the end of the string not to be parsed (the number of characters ommited was indeed equal to the number of '\n' in the string).

Let's discuss extending the test suite.

@nikhilshekhar Your work is probably affected by this PR. If you have time, please do a review too.

**Comment by Pscheidl** [comment]
@michalkurka Any specific edge cases in mind ?

- Empty row
- Extremely big data.
- Special characters inside (which ones ?)

**Comment by nikhilshekhar** [comment]
@Pscheidl FYI - I created a branch for my fix from your branch. I am making use of the `quoteCount` field that you are populating. My PR is https://github.com/h2oai/h2o-3/pull/2180. Its still WIP.

**Comment by nikhilshekhar** [comment]
@Pscheidl - There seems to be an issue with automatic detection of the types when I run the built JAR from you branch . If I build the master, it is able to infer the types well by itself for the file attached. 
Looks like the parser gets confused while guessing the header and column types now. 

Correctly parsed data from JAR built from master:
correct_parsing

Issue with parsing with default settings. If I choose the right delimiter and specify that the file has header, it parses fine. Screenshot of issue below:
part1_issue_parsing
part-2-issue-parsing

File used for testing the above is attached below:
[reviews_train.txt](https://github.com/h2oai/h2o-3/files/1832367/reviews_train.txt)

**Comment by Pscheidl** [comment]
thanks @nikhilshekhar . Will investigate. Just FYI, the data attached to the JIRA are something I did manual tests on and those work. 

It is strange that no tests discovered that (looks like all green now).

snimek obrazovky porizeny 2018-03-21 08-48-13

**Comment by Pscheidl** [comment]
@nikhilshekhar 

The separator was guessed wrongly. If you set separator correctly, it works. So the side effect is present in the guessing phase for some reason, provided it works on master. Other datasets, like common airlins_train.csv from our smalldata set work fine.

snimek obrazovky porizeny 2018-03-21 08-53-32

Here is a demonstration of correctly guessed airlines_train on my branch:

snimek obrazovky porizeny 2018-03-21 08-53-50

**Comment by Pscheidl** [comment]
Line 759 is causing the side effect:

```java
        if (CsvParser.isEOL(bits[offset]) && quoteCount % 2 == 0) break;
```

**Comment by Pscheidl** [comment]
Resolved. Single and double quotes are not counted together during guessSetup phase. It depends on ParseSetup.

The previously broken `review_train.txt` CSV posted by @nikhilshekhar now parses correctly. Result is 26319 lines, first line is header, which is correct. Manually tested.

**Comment by Pscheidl** [comment]
Some ParserTests tests failing on one single file `test_parse_mix.csv`.

Trying to resolve.

**Comment by Pscheidl** [comment]
One of the test files had clever comments with different quote styles. This was very clever. The actual chunk parsing counts with that, but not guessSetup phase.  

Quotes in comment sections are now ignored, which made the tests green locally. Let's wait for Jenkins.

File: h2o-core/src/main/java/water/parser/CsvParser.java [hunk 1]
```diff
@@ -739,14 +746,24 @@ else if (ParseUUID.isUUID(str))
     return resSetup;
   }
 
-  private static String[] getFirstLines(byte[] bits) {
+  private static String[] getFirstLines(byte[] bits, boolean singleQuotes) {
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/parser/CsvParser.java]: It seems like you are introducing a new functionality here (comments), this should be handled by a separate PR in my opinion.

Also it seems like once you set the comment to true you never unset it. Do we need this to fix PUBDEV-5394?

Comment (Pscheidl) [reply h2o-core/src/main/java/water/parser/CsvParser.java]: In my opinion we need this.

To fix 5394, the counting of quotes is required. And thiese adjustments are required for the tests to work. In this particular case, there is a `test_parse_mix` that uses both types of quotes (as apostrophe). There are other datasets like this. In the GuessSetup phase, there is no info about quotes, only the one coming from the user. the singleQuotes boolean is not a new functionality.

Comment (michalkurka) [reply h2o-core/src/main/java/water/parser/CsvParser.java]: yep, my concern was the comments (handling of '#'), it doesn't seem like the comment flag is ever reset back to false

Comment (Pscheidl) [reply h2o-core/src/main/java/water/parser/CsvParser.java]: @michalkurka That should be indeed ! With each new line ! Added and pushed !

**Comment by Pscheidl** [comment]
What I did is to reset the `#` commented line status with each new line in guessSetup phase.

**Comment by Pscheidl** [comment]
@michalkurka Removing the '#' handling breaks tests, so beware !  (Java ParserTest).

Please let's just agree that we really want it. I can then split it in a second.

**Comment by Pscheidl** [comment]
Also, this PR affects another one JIRA [PUBDEV-5320](https://0xdata.atlassian.net/browse/PUBDEV-5320)

**Comment by michalkurka** [comment]
@Pscheidl if it breaks tests then we need to keep it in one PR. What we need to do is to add tests to the current PR for handling of "#"
